File: stubs/java/src/main/java/com/beastmode/rmddd/domain/DomainBoundaries.java
package com.beastmode.rmddd.domain;

import java.util.Objects;
import java.util.Set;
import java.util.HashSet;
import java.util.Collections;

/**
 * Defines domain boundaries for a component.
 * Specifies the domain context, capabilities, and constraints.
 */
public class DomainBoundaries {
    
    private final String context;
    private final String aggregateType;
    private final Set<String> capabilities;
    private final Set<String> constraints;
    private final String ubiquitousLanguage;
    
    private DomainBoundaries(Builder builder) {
        this.context = Objects.requireNonNull(builder.context, "Context cannot be null");
        this.aggregateType = builder.aggregateType;
        this.capabilities = Collections.unmodifiableSet(new HashSet<>(builder.capabilities));
        this.constraints = Collections.unmodifiableSet(new HashSet<>(builder.constraints));
        this.ubiquitousLanguage = builder.ubiquitousLanguage;
    }
    
    public String getContext() {
        return context;
    }
    
    public String getAggregateType() {
        return aggregateType;
    }
    
    public Set<String> getCapabilities() {
        return capabilities;
    }
    
    public Set<String> getConstraints() {
        return constraints;
    }
    
    public String getUbiquitousLanguage() {
        return ubiquitousLanguage;
    }
    
    public boolean hasCapability(String capability) {
        return capabilities.contains(capability);
    }
    
    public boolean hasConstraint(String constraint) {
        return constraints.contains(constraint);
    }
    
    public static Builder builder() {
        return new Builder();
    }
    
    public static class Builder {
        private String context;
        private String aggregateType;
        private Set<String> capabilities = new HashSet<>();
        private Set<String> constraints = new HashSet<>();
        private String ubiquitousLanguage;
        
        public Builder context(String context) {
            this.context = context;
            return this;
        }
        
        public Builder aggregateType(String aggregateType) {
            this.aggregateType = aggregateType;
            return this;
        }
        
        public Builder capability(String capability) {
            this.capabilities.add(capability);
            return this;
        }
        
        public Builder capabilities(Set<String> capabilities) {
            this.capabilities.addAll(capabilities);
            return this;
        }
        
        public Builder constraint(String constraint) {
            this.constraints.add(constraint);
            return this;
        }
        
        public Builder constraints(Set<String> constraints) {
            this.constraints.addAll(constraints);
            return this;
        }
        
        public Builder ubiquitousLanguage(String ubiquitousLanguage) {
            this.ubiquitousLanguage = ubiquitousLanguage;
            return this;
        }
        
        public DomainBoundaries build() {
            return new DomainBoundaries(this);
        }
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DomainBoundaries that = (DomainBoundaries) o;
        return Objects.equals(context, that.context) &&
               Objects.equals(aggregateType, that.aggregateType) &&
               Objects.equals(capabilities, that.capabilities) &&
               Objects.equals(constraints, that.constraints) &&
               Objects.equals(ubiquitousLanguage, that.ubiquitousLanguage);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(context, aggregateType, capabilities, constraints, ubiquitousLanguage);
    }
    
    @Override
    public String toString() {
        return "DomainBoundaries{" +
               "context='" + context + '\'' +
               ", aggregateType='" + aggregateType + '\'' +
               ", capabilities=" + capabilities +
               ", constraints=" + constraints +
               ", ubiquitousLanguage='" + ubiquitousLanguage + '\'' +
               '}';
    }
}

File: stubs/csharp/src/BeastMode.RmDdd/Domain/DomainBoundaries.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace BeastMode.RmDdd.Domain
{
    /// <summary>
    /// Defines domain boundaries for a component.
    /// Specifies the domain context, capabilities, and constraints following .NET conventions.
    /// </summary>
    public class DomainBoundaries
    {
        public string Context { get; }
        public string? AggregateType { get; }
        public IReadOnlySet<string> Capabilities { get; }
        public IReadOnlySet<string> Constraints { get; }
        public string? UbiquitousLanguage { get; }
        
        private DomainBoundaries(Builder builder)
        {
            Context = builder.Context ?? throw new ArgumentNullException(nameof(builder.Context));
            AggregateType = builder.AggregateType;
            Capabilities = builder.Capabilities.ToHashSet();
            Constraints = builder.Constraints.ToHashSet();
            UbiquitousLanguage = builder.UbiquitousLanguage;
        }
        
        public bool HasCapability(string capability)
        {
            return Capabilities.Contains(capability);
        }
        
        public bool HasConstraint(string constraint)
        {
            return Constraints.Contains(constraint);
        }
        
        public static Builder Builder()
        {
            return new Builder();
        }
        
        public class Builder
        {
            internal string? Context { get; private set; }
            internal string? AggregateType { get; private set; }
            internal HashSet<string> Capabilities { get; } = new();
            internal HashSet<string> Constraints { get; } = new();
            internal string? UbiquitousLanguage { get; private set; }
            
            public Builder WithContext(string context)
            {
                Context = context;
                return this;
            }
            
            public Builder WithAggregateType(string aggregateType)
            {
                AggregateType = aggregateType;
                return this;
            }
            
            public Builder WithCapability(string capability)
            {
                Capabilities.Add(capability);
                return this;
            }
            
            public Builder WithCapabilities(IEnumerable<string> capabilities)
            {
                foreach (var capability in capabilities)
                {
                    Capabilities.Add(capability);
                }
                return this;
            }
            
            public Builder WithConstraint(string constraint)
            {
                Constraints.Add(constraint);
                return this;
            }
            
            public Builder WithConstraints(IEnumerable<string> constraints)
            {
                foreach (var constraint in constraints)
                {
                    Constraints.Add(constraint);
                }
                return this;
            }
            
            public Builder WithUbiquitousLanguage(string ubiquitousLanguage)
            {
                UbiquitousLanguage = ubiquitousLanguage;
                return this;
            }
            
            public DomainBoundaries Build()
            {
                return new DomainBoundaries(this);
            }
        }
        
        public override bool Equals(object? obj)
        {
            if (obj is not DomainBoundaries other) return false;
            if (ReferenceEquals(this, other)) return true;
            
            return Context == other.Context &&
                   AggregateType == other.AggregateType &&
                   Capabilities.SetEquals(other.Capabilities) &&
                   Constraints.SetEquals(other.Constraints) &&
                   UbiquitousLanguage == other.UbiquitousLanguage;
        }
        
        public override int GetHashCode()
        {
            return HashCode.Combine(Context, AggregateType, Capabilities, Constraints, UbiquitousLanguage);
        }
        
        public override string ToString()
        {
            return $"DomainBoundaries {{ Context = '{Context}', AggregateType = '{AggregateType}', " +
                   $"Capabilities = [{string.Join(", ", Capabilities)}], " +
                   $"Constraints = [{string.Join(", ", Constraints)}], " +
                   $"UbiquitousLanguage = '{UbiquitousLanguage}' }}";
        }
    }
}